Rebuilding speedup: check change time and checksum before syncing disk files

derekbit opened the pull request:
[Longhorn 4783](https://github.com/longhorn/longhorn/issues/4783)
    
Signed-off-by: Derek Su <derek.su@suse.com>

derekbit committed:
Unify error log messages

Longhorn 4783

Signed-off-by: Derek Su <derek.su@suse.com>

derekbit committed:
rebuilding speedup: check change time and checksum before syncing disk files

Longhorn 4783

Signed-off-by: Derek Su <derek.su@suse.com>

derekbit commented:
Runing e2e. Will rebase later.

innobead commented on sparse/rest/handlers.go:
```diff
@@ -111,37 +119,62 @@ func (server *SyncServer) doOpen(request *http.Request) error {
 	return nil
 }
 
+func (server *SyncServer) getFileIo(directIO bool) (sparse.FileIoProcessor, error) {
```
This is for creating a new instance, so it can be named `newXXXX` instead

derekbit replied:
Updated. Thank you.

innobead commented on sparse/rest/handlers.go:
```diff
@@ -111,37 +119,62 @@ func (server *SyncServer) doOpen(request *http.Request) error {
+			err = util.SetSnapshotHashInfoToChecksumFile(server.filePath+".checksum", info)
+		}
+
+		log.WithError(err).Warnf("Failed ot set snapshot hash info to checksum file %v", server.filePath)
```
Should only log error when err is not nil.

derekbit replied:
Updated. Thank you.

innobead commented on sparse/rest/handlers.go:
```diff
@@ -189,28 +222,84 @@ func (server *SyncServer) doGetChecksum(writer http.ResponseWriter, request *htt
 }
+
+func (server *SyncServer) getRecordedMetadataFromChecksumFile() ([]byte, error) {
+	f, err := os.Open(server.filePath + ".checksum")
```
`.checksum` shows twice, so suggest having this as constant.

derekbit replied:
Updated. Thank you.

innobead commented on sparse/rest/handlers.go:
```diff
@@ -111,37 +119,62 @@ func (server *SyncServer) doOpen(request *http.Request) error {
+				ChangeTime: changeTime,
+			}
+
+			err = util.SetSnapshotHashInfoToChecksumFile(server.filePath+".checksum", info)
```
So this just means after sync, save the checksum and ctime from the sender?

derekbit replied:
Yes. In the case, the disk file might be missing or mismatched with the one in the sender side. The checksum file will be created by the information from sender. Then, the rehashing is not necessary.

innobead commented on sparse/client.go:
```diff
@@ -52,80 +61,113 @@ func newHTTPClient() *http.Client {
 }
 
-func SyncContent(sourceName string, rw ReaderWriterAt, size int64, remote string, timeout int, directIO bool) (err error) {
+func getFileIo(localPath string, directIO bool) (FileIoProcessor, error) {
```
Rename to `newFileIoProcessor`

derekbit replied:
Updated. Thank you.

innobead commented on sparse/client.go:
```diff
@@ -138,14 +180,80 @@ func (client *syncClient) syncContent() error {
+	return info.Checksum, nil
+}
+
+func getLocalDiskFileChangeTimeAndChecksum(sourceName string) (string, string, string, error) {
```
nit: when having more than 2 return vars, I would suggest having named return vars to make the protocol clear.

innobead commented on sparse/client.go:
```diff
@@ -206,22 +314,25 @@ func (client *syncClient) openServer() error {
 	return nil
 }
 
 func (client *syncClient) closeServer() {
```
you renamed `openServer` to `open`, so it's better to have a consistent change here like `closeServer` to `close`.

innobead replied:
nit: I would suggest having an interface for `syncclient` to abstract the functions to map to the exposed functions of the remote sync server like `open, close, sendHole, writeData, getChecksum, getRecordedMetadata`, and then `SyncClient` can implement the interface, so we will have a good understanding for the protocol in between and good for the future unit testing. 

```go
	router.HandleFunc("/v1-ssync/open", server.open).Methods("GET")
	router.HandleFunc("/v1-ssync/close", server.close).Methods("POST")
	router.HandleFunc("/v1-ssync/sendHole", server.sendHole).Methods("POST")
	router.HandleFunc("/v1-ssync/writeData", server.writeData).Methods("POST")
	router.HandleFunc("/v1-ssync/getChecksum", server.getChecksum).Methods("GET")
	router.HandleFunc("/v1-ssync/getRecordedMetadata", server.getRecordedMetadata).Methods("GET")

```

innobead commented on util/hash.go:
```diff
@@ -0,0 +1,45 @@
+	tmpFileSuffix = ".tmp"
+)
+
+func SetSnapshotHashInfoToChecksumFile(checksumFileName string, info *types.SnapshotHashInfo) error {
```
nit: add unit test.

derekbit replied:
Addressed. Thank you.

innobead reviewed (changes requested):
LGTM, but some minor comments needed to resolve.

derekbit committed:
Add DataSyncClient and DataSyncServer interfaces

The interfaces will be used for the unit tests which will be added later.

Longhorn 4783

Signed-off-by: Derek Su <derek.su@suse.com>

innobead commented on sparse/client.go:
```diff
@@ -52,51 +61,84 @@ func newHTTPClient() *http.Client {
-	} else {
-		fileIo, err = NewBufferedFileIoProcessor(localPath, os.O_RDONLY, 0)
-	}
+	log.Infof("Syncing file %v: size %v, directIO %v, fastSync %v", localPath, fileSize, directIO, fastSync)
```
You can also add `remote` to the log

derekbit replied:
Updated, thank you.

innobead reviewed (dismissed):
LGTM

innobead commented on util/hash_test.go:
```diff
@@ -0,0 +1,33 @@
+func TestSetSnapshotHashInfoToChecksumFile(t *testing.T) {
+	assert := require.New(t)
+
+	checksumFileName := randomID(8) + types.DiskChecksumSuffix
```
nit: it's better to delete the file in teardown via defer.

derekbit replied:
Addressed. Thank you.

innobead commented on sparse/rest/server.go:
```diff
@@ -9,6 +9,15 @@ import (
 	"github.com/longhorn/sparse-tools/sparse"
 )
 
+type DataSyncServer interface {
```
This is good! 

However, because this interface is only used internally (rest pkg), so suggest renaming all methods to lower camel case.

innobead commented on sparse/client.go:
```diff
@@ -9,26 +9,44 @@ import (
+	defaultSyncWorkerCount = 4
 )
 
+type DataSyncClient interface {
```
The same, because it's used internally, so renaming methods to lower camel case intead.

derekbit replied:
Addressed. Thank you.
